File: src/frameserver/terminal/default/cli_parse.c
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <stdbool.h>
#include <stdint.h>

#include "cli_builtin.h"

/* test vector:
 * this is just a string => {this, is, just, a, string, NULL}
 * this 'can be' grouped => {this, can be, grouped, NULL}
 * this `subexp` => {this, (expand-res), NULL} */

/* unescaped presence of any of these characters enters that parsing group,
 * and when the next unescaped presence of the same character occurs, split
 * off the current work buffer into the argv array, these are just single
 * character atm, might need to do $(..) style as well */
static ssize_t find_escape_group(struct group_ent* grp, char ch)
{
	for (size_t i = 0; grp[i].enter; i++)
		if (ch == grp[i].enter)
			return i;

	return -1;
}

char** extract_argv(const char* message,
	struct argv_parse_opt opts, ssize_t* esc_ind)
{
	struct group_ent* groups = opts.groups;
	char sep = opts.sep;

/* just overfit, not worth the extra work */
	size_t len = strlen(message) + 1;
	size_t len_buf_sz = sizeof(char*) * (len + opts.prepad);
	char** argv = malloc(len_buf_sz);
	size_t arg_i = opts.prepad;
	memset(argv, '\0', len_buf_sz);

	bool esc_ign = false;

	char* work = malloc(len);
	size_t pos = 0;
	work[0] = '\0';
	*esc_ind = -1;

	for (size_t i = 0; i < len - 1; i++){
		char ch = message[i];

/* first backslash outside of a postprocessing scope, consume on use */
		if (esc_ign){
			switch (ch){
			case '\\':
				work[pos++] = ch;
			break;
			case 'n':
			 work[pos++] = '\n';
			break;
			case 't':
			 work[pos++] = '\t';
			break;
			case ' ':
			 work[pos++] = ' ';
			break;
/* re- add the backslash for non-group characters */
			default:
				if (-1 == find_escape_group(groups, ch) && ch != sep)
					work[pos++] = '\\';
				work[pos++] = ch;
			}
			esc_ign = false;
			continue;
		}

/* in escaping group? check for exit and forward or leave */
		if (-1 != *esc_ind){
			struct group_ent* group = &groups[*esc_ind];

/* constraint here is that the expansion can only yield one discrete argument
 * in expansion, if that is annoying, make argv size dynamic and have expand
 * manupulate argv and step counter manually */
			if (group->leave == ch){
				work[pos] = '\0';
				char* exp = group->expand(group, work);
				*esc_ind = -1;
				if (exp)
					argv[arg_i++] = exp;
			}

			pos = 0;
			*esc_ind = -1;
			continue;
		}

/* got one of the escape groups that might warrant different postprocessing or
 * interpretation, e.g. execute and absorb into buffer, forward to script
 * engine or other expansion */
		ssize_t ind = find_escape_group(groups, ch);

/* Design decision, should this expand 'in-str' with prefix/suffix or treat
 * as discrete entry and split off? current tactic is no - if that turns out
 * to be an annoyance, inject the prefix part into the group handler as an
 * input argument, and above where exp is created, swap that into work. This
 * requires use of [work] to track len and be able to grow beyond normal
 * string length as we know nothing of the expansion itself. */
		if (-1 != ind){
			if (pos){
				work[pos] = '\0';
				argv[arg_i++] = strdup(work);
			}

			pos = 0;
			*esc_ind = ind;
			continue;
		}

/* outside escape group, but escape next character */
		if (ch == '\\'){
			esc_ign = true;
			continue;
		}

/* finish and append to argv, ignore leading whitespace */
		if (ch == sep){
			if (pos){
				work[pos] = '\0';
				argv[arg_i++] = strdup(work);
				pos = 0;
			}
			continue;
		}

/* or append to work */
		work[pos++] = ch;
	}

/* In escape state or escape group state? parsing error */
	if (esc_ign)
		goto err_out;

/* a special attribute is required for ending while in group state,
 * this will permit things like $HOME to have an escape group of $ */
	if (*esc_ind != -1){
		struct group_ent* group = &groups[*esc_ind];

		if (!group->leave_eol)
			goto err_out;

		work[pos] = '\0';
		char* exp = group->expand(group, work);
		*esc_ind = -1;
		if (exp)
			argv[arg_i++] = exp;
	}

/* argv is already null terminated, but might have dangling arg */
	if (pos){
		work[pos] = 0;
		argv[arg_i++] = strdup(work);
	}

	free(work);
	return argv;

err_out:
	free(work);
	for (size_t i = 0; i < len; i++){
		if (argv[i])
			free(argv[i]);
	}
	free(argv);
	if (esc_ign)
		*esc_ind = len;

	return NULL;
}

File: src/frameserver/terminal/default/cli_parse.zig
// Zig reimplementation of cli_parse.c
// CLI argument parser for arcan terminal frameserver.
//
// Exports: extract_argv
//
const std = @import("std");

const shmif = @import("shmif_types");

/// Dispatch struct stitching together every C-namespace symbol the file
/// uses, routed to the right replacement module.
const c = struct {
    pub const struct_group_ent = shmif.struct_group_ent;
    pub const struct_argv_parse_opt = shmif.struct_argv_parse_opt;
};

const malloc = shmif.malloc;
const free = shmif.free;
const strdup = shmif.strdup;
const strlen = shmif.strlen;
const memset = shmif.memset;

// test vector:
//  this is just a string => {this, is, just, a, string, NULL}
//  this 'can be' grouped => {this, can be, grouped, NULL}
//  this `subexp` => {this, (expand-res), NULL}

/// unescaped presence of any of these characters enters that parsing group,
/// and when the next unescaped presence of the same character occurs, split
/// off the current work buffer into the argv array
fn find_escape_group(grp: [*c]c.struct_group_ent, ch: u8) isize {
    var i: usize = 0;
    while (grp[i].enter != 0) : (i += 1) {
        if (ch == @as(u8, @bitCast(grp[i].enter)))
            return @intCast(i);
    }
    return -1;
}

/// C char** as a Zig pointer — nullable entries (for prepad slots and NULL terminator).
/// [*c][*c]u8 in Zig: outer [*c] is many-pointer (nullable C-style), inner [*c] likewise.
export fn extract_argv(
    message: [*c]const u8,
    opts: c.struct_argv_parse_opt,
    esc_ind: *isize,
) callconv(.c) [*c][*c]u8 {
    const groups = opts.groups;
    const sep: u8 = @bitCast(opts.sep);

    // just overfit, not worth the extra work
    const len: usize = strlen(message) + 1;
    const len_buf_sz: usize = @sizeOf([*c]u8) * (len + opts.prepad);

    const argv_raw: *anyopaque = malloc(len_buf_sz) orelse return null;
    const argv: [*][*c]u8 = @ptrCast(@alignCast(argv_raw));
    _ = memset(argv_raw, 0, len_buf_sz);

    var arg_i: usize = opts.prepad;
    var esc_ign: bool = false;

    const work_raw: *anyopaque = malloc(len) orelse {
        free(argv_raw);
        return null;
    };
    const work: [*]u8 = @ptrCast(work_raw);
    var pos: usize = 0;
    work[0] = 0;
    esc_ind.* = -1;

    // Main parse loop — uses labeled block so error paths can break out.
    const success: bool = success: {
        var i: usize = 0;
        while (i < len - 1) : (i += 1) {
            const ch: u8 = message[i];

            // first backslash outside of a postprocessing scope, consume on use
            if (esc_ign) {
                switch (ch) {
                    '\\' => {
                        work[pos] = ch;
                        pos += 1;
                    },
                    'n' => {
                        work[pos] = '\n';
                        pos += 1;
                    },
                    't' => {
                        work[pos] = '\t';
                        pos += 1;
                    },
                    ' ' => {
                        work[pos] = ' ';
                        pos += 1;
                    },
                    // re-add the backslash for non-group characters
                    else => {
                        if (find_escape_group(groups, ch) == -1 and ch != sep) {
                            work[pos] = '\\';
                            pos += 1;
                        }
                        work[pos] = ch;
                        pos += 1;
                    },
                }
                esc_ign = false;
                continue;
            }

            // in escaping group? check for exit and forward or leave
            if (esc_ind.* != -1) {
                const group_idx: usize = @intCast(esc_ind.*);

                // constraint: expansion can only yield one discrete argument
                if (groups[group_idx].leave == @as(i8, @bitCast(ch))) {
                    work[pos] = 0;
                    if (groups[group_idx].expand) |exp_fn| {
                        const group_ptr: *c.struct_group_ent = @ptrCast(groups + group_idx);
                        const exp: [*c]u8 = exp_fn(group_ptr, work);
                        esc_ind.* = -1;
                        if (exp != null) {
                            argv[arg_i] = exp;
                            arg_i += 1;
                        }
                    } else {
                        esc_ind.* = -1;
                    }
                }

                pos = 0;
                esc_ind.* = -1;
                continue;
            }

            // got one of the escape groups that might warrant different
            // postprocessing or interpretation
            const ind: isize = find_escape_group(groups, ch);

            if (ind != -1) {
                if (pos != 0) {
                    work[pos] = 0;
                    argv[arg_i] = strdup(work);
                    arg_i += 1;
                }

                pos = 0;
                esc_ind.* = ind;
                continue;
            }

            // outside escape group, but escape next character
            if (ch == '\\') {
                esc_ign = true;
                continue;
            }

            // finish and append to argv, ignore leading whitespace
            if (ch == sep) {
                if (pos != 0) {
                    work[pos] = 0;
                    argv[arg_i] = strdup(work);
                    arg_i += 1;
                    pos = 0;
                }
                continue;
            }

            // or append to work
            work[pos] = ch;
            pos += 1;
        }

        // In escape state? parsing error
        if (esc_ign)
            break :success false;

        // a special attribute is required for ending while in group state,
        // this will permit things like $HOME to have an escape group of $
        if (esc_ind.* != -1) {
            const group_idx: usize = @intCast(esc_ind.*);

            if (!groups[group_idx].leave_eol)
                break :success false;

            work[pos] = 0;
            if (groups[group_idx].expand) |exp_fn| {
                const group_ptr: *c.struct_group_ent = @ptrCast(groups + group_idx);
                const exp: [*c]u8 = exp_fn(group_ptr, work);
                esc_ind.* = -1;
                if (exp != null) {
                    argv[arg_i] = exp;
                    arg_i += 1;
                }
            } else {
                esc_ind.* = -1;
            }
        }

        // argv is already null terminated, but might have dangling arg
        if (pos != 0) {
            work[pos] = 0;
            argv[arg_i] = strdup(work);
            arg_i += 1;
        }

        free(work_raw);
        break :success true;
    };

    if (!success) {
        // err_out: free work, free all argv entries, free argv
        free(work_raw);
        for (0..len) |j| {
            const entry: [*c]u8 = argv[j];
            if (entry != null)
                free(entry);
        }
        free(argv_raw);
        if (esc_ign)
            esc_ind.* = @as(isize, @intCast(len));
        return null;
    }

    return @ptrCast(argv);
}
